jupyterhub/apihandlers/shares.py

```diff
@@ -0,0 +1,598 @@
+        """Compute the REST API model for a share code"""
+        model = {
+            "server": self.server_model(share_code.spawner),
+            "scopes": share_code.scopes,
+            "id": f"sc_{share_code.id}",
+            "created_at": isoformat(share_code.created_at),
+            "expires_at": isoformat(share_code.expires_at),
+            "exchange_count": share_code.exchange_count,
+            "last_exchanged_at": isoformat(share_code.last_exchanged_at),
+        }
+        if code:
+            model["code"] = code
+            model["accept_url"] = url_concat(
+                self.hub.base_url + "accept-share", {"code": code}
+            )
+        return model
+
+    def _init_share_query(self, kind="share"):
+        """Initialize a query for Shares
+
+        before applying filters
+
+        A method so we can consolidate joins, etc.
+        """
+        if kind == "share":
+            class_ = orm.Share
+        elif kind == "code":
+            class_ = orm.ShareCode
+        else:
+            raise ValueError(
+                f"kind must be `share` or `code`, not {kind!r}"
+            )  # pragma: no cover
+
+        query = self.db.query(class_).options(
+            joinedload(class_.owner).lazyload("*"),
+            joinedload(class_.spawner).joinedload(orm.Spawner.user).lazyload("*"),
+        )
+        if kind == 'share':
+            query = query.options(
+                joinedload(class_.user).joinedload(orm.User.groups).lazyload("*"),
+                joinedload(class_.group).lazyload("*"),
+            )
+        return query
+
+    def _share_list_model(self, query, kind="share"):
+        """Finish a share query, returning the _model_"""
+        offset, limit = self.get_api_pagination()
+        if kind == "share":
+            model_method = self.share_model
+        elif kind == "code":
+            model_method = self.share_code_model
+        else:
+            raise ValueError(
+                f"kind must be `share` or `code`, not {kind!r}"
+            )  # pragma: no cover
+
+        if kind == "share":
+            class_ = orm.Share
+        elif kind == "code":
+            class_ = orm.ShareCode
+
+        total_count = query.count()
+        query = query.order_by(class_.id.asc()).offset(offset).limit(limit)
+        share_list = [model_method(share) for share in query if not share.expired]
+        return self.paginated_model(share_list, offset, limit, total_count)
+
+    def _lookup_spawner(self, user_name, server_name, raise_404=True):
+        """Lookup orm.Spawner for user_name/server_name
+
+        raise 404 if not found
+        """
+        user = self.find_user(user_name)
+        if user and server_name in user.orm_spawners:
+            return user.orm_spawners[server_name]
+        if raise_404:
+            raise web.HTTPError(404, f"No such server: {user_name}/{server_name}")
+        else:
+            return None
```

Comment:
Doesn't really matter, I just think it's clearer sometimes to have two mutually exclusive branches, rather than _implicit_ mutual exclusivity by having one terminating branch. Either is fine.